### PR TITLE
TypeChecker: Check if type exists before dereferencing it

### LINKED ISSUE #6109: Null dereference in TypeChecker
## Description

(Found by ossfuzz: 	13415)

The following contract

```
contract n{function(){var(x)=();assembly{e(x)}}}
```

triggers a null dereference here

https://github.com/ethereum/solidity/blob/1169004cec0fb256a1023ac6e67b692ccdac22d4/libsolidity/analysis/TypeChecker.cpp#L663

### PR DESCRIPTION
**Pull request by Marenz**
fixes #6109

### CODE REVIEW

File: test/libsolidity/syntaxTests/types/var_decl_val_mismatch.sol [hunk 1]
```diff
@@ -0,0 +1,25 @@
+contract n
+{
+	function()
+	{
+		// Used to cause a segfault
+		var (x,y) = (1);
+		var (z) = ();
+
+		assembly {
+			y
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/types/var_decl_val_mismatch.sol]: `mstore(y, z)` is probably better

**Comment by Marenz** [comment]
All ready

**Comment by Marenz** [comment]
I am changing this again

**Comment by Marenz** [comment]
Now ready.

File: libsolidity/analysis/TypeChecker.cpp [hunk 2]
```diff
@@ -1009,6 +1009,14 @@ bool TypeChecker::visit(VariableDeclarationStatement const& _statement)
 		}
 	}
 
+	if (valueTypes.size() != variables.size())
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp:1013]: Does it make sense to do the below in any case?

File: libsolidity/analysis/TypeChecker.cpp [hunk 3]
```diff
@@ -1009,6 +1009,14 @@ bool TypeChecker::visit(VariableDeclarationStatement const& _statement)
...
+	{
+		solAssert(m_errorReporter.hasErrors(), "Should have errors!");
+		for (auto const& var : variables)
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]:
```suggestion
		for (auto const& var: variables)
```

**Review by chriseth — CHANGES_REQUESTED** [review]

File: libsolidity/analysis/TypeChecker.cpp (hunk 2 above)
Comment (Marenz) [reply libsolidity/analysis/TypeChecker.cpp:1013]: The only problematic case is when we have more variables than values which is what we basically do here.
The other cases is already being taken care of.

I mean.. suppose it wouldn't hurt to do it unconditionally..

**Comment by axic** [comment]
Can you also put an assert into the assembly part where it crashed prior to the fix?

File: libsolidity/analysis/TypeChecker.cpp (hunk 2 above)
Comment (Marenz) [reply libsolidity/analysis/TypeChecker.cpp:1013]: I am not sure really. If this condition above is true we have already triggered an error.
We're only making it fatal here if the result of that is that a variable is left without a type.

Doing it unconditionally.. means we don't know for sure anymore if an error was already triggered for this problem so should we do BOOST_THROw or a new fatalError? 
I feel this introduces too many complications.

**Comment by Marenz** [comment]
Updated

**Review by chriseth — APPROVED** [review]

**Pull request merged**